besu/src/test/java/org/hyperledger/besu/controller/QbftBesuControllerBuilderTest.java

```diff
@@ -0,0 +1,169 @@
+    when(genesisConfigFile.getConfigOptions(any())).thenReturn(genesisConfigOptions);
+    when(storageProvider.createBlockchainStorage(any()))
+        .thenReturn(
+            new KeyValueStoragePrefixedKeyBlockchainStorage(
+                new InMemoryKeyValueStorage(), new MainnetBlockHeaderFunctions()));
+    when(storageProvider.createWorldStateStorage(DataStorageFormat.FOREST))
+        .thenReturn(worldStateStorage);
+    when(worldStateStorage.isWorldStateAvailable(any(), any())).thenReturn(true);
+    when(worldStateStorage.updater()).thenReturn(mock(WorldStateStorage.Updater.class));
+    when(worldStatePreimageStorage.updater())
+        .thenReturn(mock(WorldStatePreimageStorage.Updater.class));
+    when(storageProvider.createWorldStatePreimageStorage()).thenReturn(worldStatePreimageStorage);
+    when(synchronizerConfiguration.getDownloaderParallelism()).thenReturn(1);
+    when(synchronizerConfiguration.getTransactionsParallelism()).thenReturn(1);
+    when(synchronizerConfiguration.getComputationParallelism()).thenReturn(1);
+    when(observableMetricsSystem.createLabelledCounter(
+            any(), anyString(), anyString(), anyString()))
+        .thenReturn(labels -> null);
+    when(synchronizerConfiguration.getBlockPropagationRange()).thenReturn(Range.closed(1L, 2L));
+
+    // qbft prepForBuild setup
+    when(genesisConfigOptions.getQbftConfigOptions())
+        .thenReturn(new MutableQbftConfigOptions(JsonQbftConfigOptions.DEFAULT));
+    when(genesisConfigOptions.getTransitions()).thenReturn(mock(TransitionsConfigOptions.class));
+    when(genesisConfigFile.getExtraData())
+        .thenReturn(
+            QbftExtraDataCodec.createGenesisExtraDataString(List.of(Address.fromHexString("1"))));
+
+    qbftBesuControllerBuilder =
+        new QbftBesuControllerBuilder()
+            .genesisConfigFile(genesisConfigFile)
+            .synchronizerConfiguration(synchronizerConfiguration)
+            .ethProtocolConfiguration(ethProtocolConfiguration)
+            .networkId(networkId)
+            .miningParameters(miningParameters)
+            .metricsSystem(observableMetricsSystem)
+            .privacyParameters(privacyParameters)
+            .dataDirectory(tempDirRule.getRoot().toPath())
+            .clock(clock)
+            .transactionPoolConfiguration(poolConfiguration)
+            .nodeKey(nodeKey)
+            .storageProvider(storageProvider)
+            .gasLimitCalculator(gasLimitCalculator)
+            .evmConfiguration(EvmConfiguration.DEFAULT);
+  }
+
+  @Test
+  public void forkingValidatorProviderIsAvailableOnBftContext() {
+    BesuController besuController = qbftBesuControllerBuilder.build();
+
+    ValidatorProvider validatorProvider =
+        besuController
+            .getProtocolContext()
+            .getConsensusState(BftContext.class)
+            .getValidatorProvider();
+    assertThat(validatorProvider).isInstanceOf(ForkingValidatorProvider.class);
+  }
+
```

Comment:
Is it possible to add a test that shows that the QbftJsonRpcMethods are actually created with a read only validator provider?